rxgroups-android/src/main/java/com/airbnb/rxgroups/android/GroupLifecycleManager.java

```diff
@@ -0,0 +1,172 @@
+import rx.Observer;
+import rx.Subscription;
+import rx.functions.Action1;
+import rx.functions.Func1;
+import rx.subscriptions.CompositeSubscription;
+
+public class GroupLifecycleManager {
+  private static final String TAG = "GroupLifecycleManager";
+  private static final String KEY_MANAGER_HASHCODE = "KEY_MANAGER_HASHCODE";
+  private static final String KEY_GROUP_ID = "KEY_GROUP_ID";
+
+  private final CompositeSubscription pendingSubscriptions = new CompositeSubscription();
+  private final LifecycleResubscription resubscription;
+  private final ObservableManager observableManager;
+  private final ObservableTagFactory tagFactory;
+  private ObservableGroup group;
+  private boolean hasSavedState;
+
+  public GroupLifecycleManager(ObservableManager observableManager,
+      ObservableTagFactory tagFactory, LifecycleResubscription resubscription) {
+    this.observableManager = observableManager;
+    this.tagFactory = tagFactory;
+    this.resubscription = resubscription;
+    group = observableManager.newGroup();
+    group.lock();
+  }
+
+  public GroupLifecycleManager(ObservableManager observableManager,
+      ObservableTagFactory tagFactory) {
+    this(observableManager, tagFactory, new LifecycleResubscription());
+  }
+
+  public void saveState(Bundle outState) {
+    hasSavedState = true;
+    outState.putInt(KEY_MANAGER_HASHCODE, observableManager.hashCode());
+    outState.putLong(KEY_GROUP_ID, group.id());
+  }
+
+  public void restoreState(@Nullable Bundle savedState, @Nullable Object target) {
+    // Only need to resubscribe if restoring from saved state
+    boolean shouldResubscribe = target != null && savedState != null;
+
+    if (savedState != null) {
+      int hashCode = savedState.getInt(KEY_MANAGER_HASHCODE);
+      long groupId = savedState.getLong(KEY_GROUP_ID);
+
+      if (hashCode != observableManager.hashCode()) {
+        group = observableManager.newGroup();
+      } else {
+        ObservableGroup group = observableManager.getGroup(groupId);
+        if (group.isDestroyed()) {
+          Log.w(TAG, "Tried to reuse GroupLifecycleManager with a destroyed group");
+          group = observableManager.newGroup();
+          shouldResubscribe = false;
+        }
+        this.group = group;
+      }
+    }
+
+    if (shouldResubscribe) {
+      subscribe(target);
+    }
+  }
+
+  /**
+   * Returns whether the provided {@link Class} exists for the {@link ObservableGroup}.
+   * Observables will only be removed from their respective groups once {@link
+   * Observer#onCompleted()} has been called.
+   */
+  public boolean hasObservable(Class<?> klass) {
+    return hasObservable(tagFactory.tag(klass));
+  }
+
+  public boolean hasObservable(String tag) {
+    return group.hasObservable(tag);
+  }
+
+  /**
+   * Subscribe all Observer fields on the target that are annotated with {@link AutoResubscribe}
+   * and that have their corresponding Observable in flight.
+   */
+  public void subscribe(Object target) {
```

Comment:
I've never liked the confusing naming between `subscribe` and `resubscribe` (my bad on that! :P). Maybe we can thinking of something better.